consensus/ibft/consensus_backend.go

```diff
@@ -127,209 +125,46 @@ func (i *backendIBFT) Quorum(blockNumber uint64) uint64 {
-	if err != nil {
-		return nil, err
-	}
-
-	txs := i.writeTransactions(gasLimit, header.Number, transition)
-
-	if err := i.PreCommitState(header, transition); err != nil {
-		return nil, err
-	}
-
-	_, root := transition.Commit()
-	header.StateRoot = root
-	header.GasUsed = transition.TotalGas()
-
-	// build the block
-	block := consensus.BuildBlock(consensus.BuildBlockParams{
-		Header:   header,
-		Txns:     txs,
-		Receipts: transition.Receipts(),
-	})
-
-	// write the seal of the block after all the fields are completed
-	header, err = i.currentSigner.WriteProposerSeal(header)
-	if err != nil {
-		return nil, err
-	}
-
-	block.Header = header
-
-	// compute the hash, this is only a provisional hash since the final one
-	// is sealed after all the committed seals
-	block.Header.ComputeHash()
-
-	i.logger.Info("build block", "number", header.Number, "txs", len(txs))
-
-	return block, nil
-}
-
-type status uint8
-
-const (
-	success status = iota
-	fail
-	skip
-)
-
-type txExeResult struct {
-	tx     *types.Transaction
-	status status
-}
-
-type transitionInterface interface {
-	Write(txn *types.Transaction) error
-	WriteFailedReceipt(txn *types.Transaction) error
-}
-
-func (i *backendIBFT) writeTransactions(
-	gasLimit,
-	blockNumber uint64,
-	transition transitionInterface,
-) (executed []*types.Transaction) {
-	executed = make([]*types.Transaction, 0)
-
-	if !i.currentHooks.ShouldWriteTransactions(blockNumber) {
-		return
-	}
-
-	var (
-		blockTimer = time.NewTimer(i.blockTime)
-
-		successful = 0
-		failed     = 0
-		skipped    = 0
-	)
-
-	defer func() {
-		i.logger.Info(
-			"executed txs",
-			"successful", successful,
-			"failed", failed,
-			"skipped", skipped,
-			"remaining", i.txpool.Length(),
-		)
-	}()
-
-	i.txpool.Prepare()
-
-write:
-	for {
-		select {
-		case <-blockTimer.C:
-			return
-		default:
-			// execute transactions one by one
-			result, ok := i.writeTransaction(
-				i.txpool.Peek(),
-				transition,
-				gasLimit,
-			)
-
-			if !ok {
-				break write
-			}
-
-			tx := result.tx
-
-			switch result.status {
-			case success:
-				executed = append(executed, tx)
-				successful++
-			case fail:
-				failed++
-			case skip:
-				skipped++
-			}
-		}
-	}
-
-	//	wait for the timer to expire
-	<-blockTimer.C
-
-	return
-}
-
-func (i *backendIBFT) writeTransaction(
-	tx *types.Transaction,
-	transition transitionInterface,
-	gasLimit uint64,
-) (*txExeResult, bool) {
-	if tx == nil {
-		return nil, false
-	}
-
-	if tx.ExceedsBlockGasLimit(gasLimit) {
-		i.txpool.Drop(tx)
+	// fill with transactions
+	bbuilder.Fill()
 
-		if err := transition.WriteFailedReceipt(tx); err != nil {
-			i.logger.Error(
-				fmt.Sprintf(
-					"unable to write failed receipt for transaction %s",
-					tx.Hash,
-				),
-			)
+	stateBlock := bbuilder.Build(func(header *types.Header) {
+		if err := i.currentHooks.ModifyHeader(header, i.currentSigner.Address()); err != nil {
+			panic(err)
```

Comment:
Maybe we should change `Build` signature to
`func (b *BlockBuilder) Build(handler func(*types.Header) error) (*StateBlock, error) ` and get rid of panic